### ISSUE TITLE
Propagate more hints through PromQL Select

### ISSUE THREAD

**Issue by bwplotka** [prometheus/prometheus#10041]
I would love to propose pushing more hints to storage.

#### Proposal

Let's take an example query: `avg(sum(rate(abc{}[5m])) by (pod)) by (cluster)`

Currently such query will ask storage using `Querier` interface `Select` method (`Select(sortSeries bool, hints *SelectHints, matchers ...*labels.Matcher) SeriesSet`)  with `__name__ = abc` matcher and following select hints:

```
SelectHints{
	Start: ..., End: ...,  Step: ...,
	Func: "rate",
	Range: 5 * 60 * 1000,
}
```

I would love to propose to pass all hints that are scoped for given matchers FROM the deepest. So after the change storage would have something like:

```
[]SelectHints{
{ 
        Start: ..., End: ...,  Step: ...,
	Func: "rate",
	Range: 5 * 60 * 1000,
}. {
	Func: "sum"
        Grouping []string{"pod"},
	By: true.
}, {
	Func: "avg"
        Grouping []string{"cluster"},
	By: true.
}
```

### Motivation

Similar to https://github.com/prometheus/prometheus/issues/10040 such change will enormously improve how in Thanos we can improve query evaluation, how much we can pushdown to leafs and how much we can parallelize the query. 

In theory we could parallize such query in Prometheus too and perform sub PromQL eval on disjont blocks too for safe query concurrency. This is however outside of this proposal. What we want here is just to pass more information (select hints) to storage and remote read counterparts.

**Comment by brian-brazil** [comment]
That avg hint is not safe to send, as it's impossible to do things with it correctly in the general case. Thus far the only hints sent have been for things where it is possible to do correct things with it, so that sum hint is fine for example as sum is associative. min/max are also safe.

**Comment by moadz** [comment]
@brian-brazil Average can be associative if you return the sum/count, correct?

**Comment by brian-brazil** [comment]
> so that sum hint is fine for example as sum is associative.

I should correct myself here, this is not safe as different remote read endpoints could return overlapping data so you could end up over-counting.

> Average can be associative if you pushdown(up) the sum/count, correct?

No as a) you can only return one value and b) the above reason.

Min/max remain safe I believe.

**Comment by roidelapluie** [comment]
Is this a duplicate of #9979 ?

**Comment by bwplotka** [comment]
Thanks for feedback!

#9979 is connected, but not the same. #9978 is about extending hints with more parameters from different functions like `top`. Issue #10041 is about passing a slice of hints to have info about all nested aggr/func in play.

**Comment by beorn7** [comment]
Hello from the bug scrub!

@bwplotka given that we closed #9979 and #9978, we were wondering how relevant this one still is. Could you either close or comment what the way forward would be?

**Comment by bwplotka** [comment]
Thanks! 

I will let others who work on storage or engines these days to reopen/create new one with the fresh requirements if needed. 👍
